File: ai_companions_flutter/lib/core/services/premium_service.dart
// Premium Service - Coins, Subscriptions, Image Generation

import 'package:dio/dio.dart';
import '../../core/network/dio_client.dart';
import '../../shared/models/index.dart';

class PremiumService {
  PremiumService._();

  static final instance = PremiumService._();

  Dio get _dio => DioClient.instance.dio;

  // Coins
  Future<int> getBalance() async {
    try {
      final response = await _dio.get('/premium/balance');
      final data = response.data['data'] as Map<String, dynamic>;
      return data['balance'] as int;
    } catch (e) {
      rethrow;
    }
  }

  Future<CoinTransaction> claimDailyBonus() async {
    try {
      final response = await _dio.post('/premium/daily-bonus');
      final data = response.data['data'] as Map<String, dynamic>;
      return CoinTransaction.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }

  Future<PaginatedResponse<CoinTransaction>> getTransactions({int page = 1, int limit = 20}) async {
    try {
      final response = await _dio.get(
        '/premium/transactions',
        queryParameters: {'page': page, 'limit': limit},
      );
      return PaginatedResponse.fromJson(
        response.data as Map<String, dynamic>,
        (json) => CoinTransaction.fromJson(json as Map<String, dynamic>),
      );
    } catch (e) {
      rethrow;
    }
  }

  // Subscriptions
  Future<Subscription> getSubscription() async {
    try {
      final response = await _dio.get('/premium/subscription');
      final data = response.data['data'] as Map<String, dynamic>;
      return Subscription.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }

  Future<SubscriptionResponse> subscribe(String plan) async {
    try {
      final response = await _dio.post(
        '/premium/subscribe',
        data: {'plan': plan},
      );
      final data = response.data['data'] as Map<String, dynamic>;
      return SubscriptionResponse.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }

  Future<void> cancelSubscription() async {
    try {
      await _dio.post('/premium/cancel-subscription');
    } catch (e) {
      rethrow;
    }
  }

  // Image Generation
  Future<GenerationRequest> generateImage(String prompt, {String? style}) async {
    try {
      final response = await _dio.post(
        '/premium/generate',
        data: {
          'prompt': prompt,
          'style': style,
        },
      );
      final data = response.data['data'] as Map<String, dynamic>;
      return GenerationRequest.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }

  Future<PaginatedResponse<GenerationRequest>> getGenerations({int page = 1, int limit = 20}) async {
    try {
      final response = await _dio.get(
        '/premium/generations',
        queryParameters: {'page': page, 'limit': limit},
      );
      return PaginatedResponse.fromJson(
        response.data as Map<String, dynamic>,
        (json) => GenerationRequest.fromJson(json as Map<String, dynamic>),
      );
    } catch (e) {
      rethrow;
    }
  }

  Future<GenerationRequest> getGeneration(String id) async {
    try {
      final response = await _dio.get('/premium/generations/$id');
      final data = response.data['data'] as Map<String, dynamic>;
      return GenerationRequest.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }
}

File: app/src/services/premium.service.ts
import { api } from './api';
import type { CoinTransaction } from '@ai-companions/shared';

export const premiumService = {
  // ---- Coins ----
  async getBalance(): Promise<number> {
    const response = await api.get('/premium/balance');
    return response.data.data.balance;
  },

  async claimDailyBonus(): Promise<CoinTransaction> {
    const response = await api.post('/premium/daily-bonus');
    return response.data.data;
  },

  async getTransactions(page: number = 1, limit: number = 20) {
    const response = await api.get('/premium/transactions', { params: { page, limit } });
    return response.data.data;
  },

  // ---- Subscriptions ----
  async getSubscription() {
    const response = await api.get('/premium/subscription');
    return response.data.data;
  },

  async subscribe(plan: 'starter' | 'pro' | 'ultimate') {
    const response = await api.post('/premium/subscribe', { plan });
    return response.data.data;
  },

  async cancelSubscription(): Promise<void> {
    await api.post('/premium/cancel-subscription');
  },

  // ---- Image Generation ----
  async generateImage(prompt: string, style?: string) {
    const response = await api.post('/premium/generate', { prompt, style });
    return response.data.data;
  },

  async getGenerations(page: number = 1, limit: number = 20) {
    const response = await api.get('/premium/generations', { params: { page, limit } });
    return response.data.data;
  },

  async getGeneration(id: string) {
    const response = await api.get(`/premium/generations/${id}`);
    return response.data.data;
  },
};
